### PR TITLE
apply most patches in chw's repository

### LINKED ISSUE #25: TclCurl curl::versioninfo core dump, buffer overflow in curlVersion()
With tclcurl in Ubuntu 22.04.03 LTS, simply calling curl::version immediately core dumps, with a buffer overflow in __vsprintf_internal and curlVersion.  Stack trace and other details below.  This did not happen in earlier versions of Ubuntu, neither 20.04 nor 18.04.  I have not tested on any other systems.

This is a really easy bug to reproduce with TclCurl, but I do not know whether the real problem is in TclCurl, libcurl, or something specific to how Ubuntu or Debian builds those libraries.  I looked for Ubuntu and Debian tclcurl bug trackers, and could not find any that seem to actually be in use, instead of abandoned and ignored.  Thus I'm reporting it here.

Reproduce the problem:
```
$ /usr/bin/tclsh8.6
% package require TclCurl
7.22.0
% package ifneeded TclCurl 7.22.0
load /usr/lib/tcltk/TclCurl7.22.0/libTclCurl7.22.0.so
source /usr/lib/tcltk/TclCurl7.22.0/tclcurl.tcl
% curl::versioninfo -features
IPV6 SSL LIBZ NTLM ASYNCHDNS SPNEGO LARGEFILE IDN
% curl::versioninfo -protocols
dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp

% curl::version
*** buffer overflow detected ***: terminated
Aborted (core dumped)

$ gdb -c ~/core.1657856 -e /usr/bin/tclsh8.6
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=139884358782464) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=139884358782464) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=139884358782464, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007f395d642476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007f395d6287f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007f395d689676 in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7f395d7db92e "*** %s ***: terminated\n") at ../sysdeps/posix/libc_fatal.c:155
#6  0x00007f395d7363aa in __GI___fortify_fail (msg=msg@entry=0x7f395d7db8d4 "buffer overflow detected") at ./debug/fortify_fail.c:26
#7  0x00007f395d734d26 in __GI___chk_fail () at ./debug/chk_fail.c:28
#8  0x00007f395d6818df in _IO_str_chk_overflow (fp=<optimized out>, c=<optimized out>) at ./libio/iovsprintf.c:35
#9  0x00007f395d68de34 in __GI__IO_default_xsputn (n=<optimized out>, data=<optimized out>, f=<optimized out>) at ./libio/genops.c:399
#10 __GI__IO_default_xsputn (f=0x7fff2cd296a0, data=<optimized out>, n=177) at ./libio/genops.c:370
#11 0x00007f395d67700c in outstring_func (done=24, length=177, 
    string=0x7f395d4c4060 "libcurl/7.81.0 GnuTLS/3.7.3 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.16", 
    s=0x7fff2cd296a0) at ../libio/libioP.h:947
#12 __vfprintf_internal (s=s@entry=0x7fff2cd296a0, format=format@entry=0x7f395da21007 "TclCurl Version %s (%s)", ap=ap@entry=0x7fff2cd297e0, mode_flags=mode_flags@entry=6) at ./stdio-common/vfprintf-internal.c:1517
#13 0x00007f395d681989 in __vsprintf_internal (
    string=0x7fff2cd298c0 "TclCurl Version 7.22.0 (libcurl/7.81.0 GnuTLS/3.7.3 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.", maxlen=<optimized out>, format=0x7f395da21007 "TclCurl Version %s (%s)", args=args@entry=0x7fff2cd297e0, mode_flags=6) at ./libio/iovsprintf.c:95
#14 0x00007f395d734841 in ___sprintf_chk (s=<optimized out>, flag=<optimized out>, slen=<optimized out>, format=<optimized out>) at ./debug/sprintf_chk.c:40
#15 0x00007f395da189d2 in curlVersion () from /usr/lib/tcltk/TclCurl7.22.0/libTclCurl7.22.0.so
#16 0x00007f395d896d32 in TclNRRunCallbacks () from /lib/x86_64-linux-gnu/libtcl8.6.so
#17 0x00007f395d924ab4 in Tcl_RecordAndEvalObj () from /lib/x86_64-linux-gnu/libtcl8.6.so
#18 0x00007f395d94b0a7 in Tcl_MainEx () from /lib/x86_64-linux-gnu/libtcl8.6.so
#19 0x000055a953909109 in ?? ()
#20 0x0000000000000000 in ?? ()
(gdb) exit
```

More info on the software versions where the crash occurred:
```
$ grep DISTRIB_DESCRIPTION /etc/lsb-release
DISTRIB_DESCRIPTION="Ubuntu 22.04.3 LTS"
$ /usr/bin/curl --version
curl 7.81.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.16
Release-Date: 2022-01-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd

$ dpkg -l 'libcurl*'
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                  Version            Architecture Description
+++-=====================-==================-============-==============================================================
un  libcurl3              <none>             <none>       (no description available)
ii  libcurl3-gnutls:amd64 7.81.0-1ubuntu1.15 amd64        easy-to-use client-side URL transfer library (GnuTLS flavour)
un  libcurl3-nss          <none>             <none>       (no description available)
ii  libcurl4:amd64        7.81.0-1ubuntu1.15 amd64        easy-to-use client-side URL transfer library (OpenSSL flavour)
un  libcurl4-gnutls-dev   <none>             <none>       (no description available)

$ dpkg -S /usr/lib/tcltk/TclCurl7.22.0/libTclCurl7.22.0.so
tclcurl: /usr/lib/tcltk/TclCurl7.22.0/libTclCurl7.22.0.so
$ dpkg -S /usr/bin/tclsh8.6
tcl8.6: /usr/bin/tclsh8.6
$ dpkg -S /usr/bin/curl
curl: /usr/bin/curl
$ apt show tcl8.6 | grep Version
Version: 8.6.12+dfsg-1build1
$ apt show tclcurl | grep Version
Version: 7.22.0+hg20160822-3
$ apt show curl | grep Version
Version: 7.81.0-1ubuntu1.15
$ apt show libcurl4:amd64 | grep Version
Version: 7.81.0-1ubuntu1.15
```

### PR DESCRIPTION
**Pull request by mxmanghi**
I applied to the current FA code for tclcurl the changes made in androwish's repository  (https://www.androwish.org/index.html/tktview/b7f1f6c72d7026ef04b7e77b687de25a7cc836d7). I also reworked the curlWriteProcInvoke function that caused Tcl to segfault (!) when the call back definition was made by multiple words, like in the case where the callback is implemented by an object method.

### CODE REVIEW

**Comment by resuna** [comment]
Why was this closed?

**Comment by mxmanghi** [comment]
my mistake, I'm going to reopen it

  -- M

**Comment by resuna** [comment]
Thank you!

**Comment by resuna** [comment]
Looks like it would benefit from having enums matching the method/option lists in tclcurl.h, it would make tclcurl.c easier to follow. :)

**Comment by bovine** [comment]
Do you have a test case that can reproduce the crash?

**Comment by mxmanghi** [comment]
Yes definitely

  -- M

**Comment by mxmanghi** [comment]
> Do you have a test case that can reproduce the crash?

I have 3 cases. They have been tested on Ubuntu Noble, but also with a custom installation of Tcl 8.6.14 with optimizer disabled and symbols loaded

```
1)
tclsh8.6 [~/Documents/tclcurl] package require TclCurl
7.22.0
tclsh8.6 [~/Documents/tclcurl] set curl [::curl::init]
curl1
tclsh8.6 [~/Documents/tclcurl] $curl configure -url https://centromajorana.it/ -header 1 -timeout 5000 -writeproc [namespace current]::invalidproc
tclsh8.6 [~/Documents/tclcurl] $curl perform
Segmentation fault (core dumped)

2)
tclsh8.6 [~/Documents/tclcurl] package require TclCurl
7.22.0
tclsh8.6 [~/Documents/tclcurl] set curl [::curl::init]
curl1
tclsh8.6 [~/Documents/tclcurl] ::oo::class create CurlHandler
::CurlHandler
tclsh8.6 [~/Documents/tclcurl] ::oo::define CurlHandler { method callback {s} { puts $s } }
tclsh8.6 [~/Documents/tclcurl] set ch [CurlHandler new]
tclsh8.6 [~/Documents/tclcurl] $curl configure -url https://centromajorana.it/ -header 1 -timeout 5000 -writeproc [list $ch callback]
tclsh8.6 [~/Documents/tclcurl] $curl perform
Segmentation fault (core dumped)

3)
tclsh8.6 [~/Documents/tclcurl] package require TclCurl
7.22.0
tclsh8.6 [~/Documents/tclcurl] proc callback {s} { puts [string length $s] }
tclsh8.6 [~/Documents/tclcurl] set curl [::curl::init]
curl1
tclsh8.6 [~/Documents/tclcurl] $curl configure -url https://centromajorana.it/ -header 1 -timeout 5000 -writeproc callback
tclsh8.6 [~/Documents/tclcurl] $curl perform
17
37
32
67
23
28
40
2
7940
1663
0
tclsh8.6 [~/Documents/tclcurl] 
```

**Comment by bovine** [comment]
If you sync your branch with our master, the broken CI builds should be fixed here too.

**Comment by bovine** [comment]
the Mac CI is legitimately failing on something involving `SetoptBlob` not being defined.

In the `curlVersionInfo` function, the `tmp[7]` variable can be removed since its only use was eliminated by the use of Tcl_ObjPrintf in the `case 1` block.

There are a couple of other warnings as well.

File: generic/tclcurl.c [hunk 1]
```diff
@@ -87,27 +93,26 @@ Tclcurl_Init (Tcl_Interp *interp) {
...
-    char                *handleName;
+    char                handleName[32];
     int                 i;
     Tcl_CmdInfo         info;
     Tcl_Command         cmdToken;
 
     /* We try with curl1, if it already exists with curl2...*/
-    handleName=(char *)Tcl_Alloc(32);
     for (i=1;;i++) {
         sprintf(handleName,"curl%d",i);
```
Comment (bovine) [inline generic/tclcurl.c]:
```suggestion
        snprintf(handleName,sizeof(handleName),"curl%d",i);
```

File: generic/tclcurl.c [hunk 2]
```diff
@@ -4220,28 +4365,26 @@ char
...
-    char                *shandleName;
+    char                shandleName[32];
     int                 i;
     Tcl_CmdInfo         info;
     Tcl_Command         cmdToken;
 
     /* We try with scurl1, if it already exists with scurl2...*/
-    shandleName=(char *)Tcl_Alloc(32);
     for (i=1;;i++) {
         sprintf(shandleName,"scurl%d",i);
```
Comment (bovine) [inline generic/tclcurl.c]:
```suggestion
        snprintf(shandleName,sizeof(shandleName),"scurl%d",i);
```

**Review by bovine — APPROVED** [review]
Seems like fine improvements. I've suggested a couple of places where snprintf could be used during the handle generation. 

@resuna might want to review the entire PR also.

File: generic/tclcurl.c [hunk 3]
```diff
@@ -2847,25 +2992,27 @@ curlChunkEndProcInvoke (void *curlDataPtr) {
...
 
     struct curlObjData      *curlData=(struct curlObjData *)curlDataPtr;
     Tcl_Obj                 *tclProcPtr;
-    char                     tclCommand[500];
     int                      i;
 
-    snprintf(tclCommand,500,"%s %s %s",curlData->fnmatchProc,pattern,filename);
-    tclProcPtr=Tcl_NewStringObj(tclCommand,-1);
-
+    tclProcPtr=Tcl_ObjPrintf("%s %s %s",curlData->fnmatchProc,pattern,filename);
```
Comment (resuna) [inline generic/tclcurl.c:3009]: Should this be built up as a list? What if pattern contains spaces?

File: generic/tclcurl.h [hunk 4]
```diff
@@ -239,11 +243,11 @@ CONST static char *configTable[] = {
...
-    "-sourceprequote",    "-sourcepostquote",    "-ftpsslauth",
-    "-sourceurl",         "-sourcequote",        "-ftpaccount",
-    "-ignorecontentlength",                      "-cookielist",
-    "-ftpskippasvip",     "-ftpfilemethod",      "-localport",
-    "-localportrange",
+    "-sourceprequote",    "-sourcepostquote",
+    "-ftpsslauth",        "-sourceurl",          "-sourcequote",
+    "-ftpaccount",        "-ignorecontentlength",
+    "-cookielist",        "-ftpskippasvip",
+    "-ftpfilemethod",     "-localport",          "-localportrange",
```
Comment (resuna) [inline generic/tclcurl.h:250]: Was anything actually changed here?

File: generic/tclcurl.h [hunk 5]
```diff
@@ -488,13 +495,14 @@ void curlSetBodyVarName(Tcl_Interp *interp,struct curlObjData *curlDataPtr);
...
 
-char *curlCreateShareObjCmd (Tcl_Interp *interp,struct shcurlObjData  *shcurlData);
+Tcl_Obj* curlCreateShareObjCmd (Tcl_Interp *interp,struct shcurlObjData  *shcurlData);
 int curlShareInitObjCmd (ClientData clientData, Tcl_Interp *interp,
         int objc,Tcl_Obj *CONST objv[]);
 int curlShareObjCmd (ClientData clientData, Tcl_Interp *interp,
         int objc,Tcl_Obj *CONST objv[]);
 int curlCleanUpShareCmd(ClientData clientData);
 
+#ifndef multi_h
```
Comment (resuna) [inline generic/tclcurl.h:505]: Is there any possibility that multi_h won't be set? multi.c (and thus multi.h) seem to be incorporated unconditionally in the config files.

**Review by resuna — CHANGES_REQUESTED** [review]

File: generic/tclcurl.h (hunk 4 above)
Comment (mxmanghi) [reply generic/tclcurl.h:250]: No, it's easy to verify that the 2 option lists are just reorganized

**Comment by mxmanghi** [comment]
I applied the last change suggested by Jeff. What is your policy for releasing? Is it worth a release? I'm making out of the whole stuff a package for Debian/Ubuntu, I can apply the new improvements as patches for the Debian packaging system but I rather prefer to release as is with a new patchlevel number

**Comment by bovine** [comment]
after we merge this PR, we might want to also get PR 22 (I suspect there will be some minor merge conflicts), and then tag a new release.

**Comment by bovine** [comment]
I've tagged v7.22.1 and left that other PR unmerged for now. It can be addressed in the future.

**Comment by mxmanghi** [comment]
Great! Thank you

**Pull request merged**